File: LeetCodeExample.Test/Binary Tree/_00112_PathSum.cs
using LeetCodeExample.Test.Common;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test
{
    /// <summary>
    //Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.
    //A leaf is a node with no children.
    /// </summary>
    public class _00112_PathSum
    {
        public bool HasPathSum(TreeNode root, int targetSum)
        {
            return Recurse(root, 0, targetSum);
        }

        bool Recurse(TreeNode root, int currentSum, int targetSum)
        {
            if (root == null)
                return false;

            // Reached a leaf, check sums
            if (root.left == null && root.right == null)
                return currentSum + root.val == targetSum;

            if (Recurse(root.left, currentSum + root.val, targetSum))
                return true;
            return Recurse(root.right, currentSum + root.val, targetSum);
        }
    }
}

File: LeetCodeExample.Test/Binary Tree/_00112_PathSum.py
# Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.
# A leaf is a node with no children.
# Definition for a binary tree node.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import Optional
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def rec(node: Optional[TreeNode], val: int) -> bool:
            if not node:
                return False
            node_val = node.val + val
            if node.left == node.right == None:
                return targetSum == node_val
            return rec(node.left, node_val) or rec(node.right, node_val)
        return rec(root, 0)
    


# Old try
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def rec(node: Optional[TreeNode], sum: int) -> bool:
            if not node:
                return False
            sum += node.val
            # Ensure to only check on a leaf node
            if not node.left and not node.right and sum == targetSum:
                return True
            if rec(node.left, sum):
                return True
            return rec(node.right, sum)
        return rec(root, 0)